### PR TITLE
feat(db): switch from settype to casts

### PR DESCRIPTION
**Pull request by miaulalala**
Signed-off-by: Anna Larch <anna@nextcloud.com>

## Summary
casting directly to the proper type is up to 6 times faster than using `settype`

Also added the types for the methods, let's see if things go boom (although they might not without strict types)

## TODO

- [ ] ...

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

File: lib/public/AppFramework/Db/Entity.php [hunk 1]
```diff
@@ -76,50 +78,63 @@ public function getFieldTypes() {
...
-				$type = $this->_fieldTypes[$name];
-				if ($type === 'blob') {
-					// (B)LOB is treated as string when we read from the DB
-					if (is_resource($args[0])) {
-						$args[0] = stream_get_contents($args[0]);
-					}
-					$type = 'string';
+		if (!property_exists($this, $name)) {
+			throw new \BadFunctionCallException($name .
+				' is not a valid attribute');
```
Comment (provokateurin) [inline lib/public/AppFramework/Db/Entity.php:96]:
```suggestion
			throw new \BadFunctionCallException($name . ' is not a valid attribute');
```

File: lib/public/AppFramework/Db/Entity.php [hunk 2]
```diff
@@ -76,50 +78,63 @@ public function getFieldTypes() {
...
+		}
+
+		if ($args[0] === $this->$name) {
+			return;
+		}
```
Comment (provokateurin) [inline lib/public/AppFramework/Db/Entity.php:101]: I still don't understand the purpose of this return :thinking:

**Review by provokateurin — APPROVED** [review]
Noice!

File: lib/public/AppFramework/Db/Entity.php (hunk 2 above)
Comment (miaulalala) [reply lib/public/AppFramework/Db/Entity.php:101]: It was already there and I don't wanna touch it :scream: 

(I think it has to do with marking a column as updated or not)

Comment (provokateurin) [reply lib/public/AppFramework/Db/Entity.php:100]: I know ;)

Comment (kesselb) [reply lib/public/AppFramework/Db/Entity.php:100]: Do nothing if the current value is equal to the new one

**Review by AndyScherzinger — APPROVED** [review]

File: lib/public/AppFramework/Db/Entity.php (hunk 2 above)
Comment (provokateurin) [reply lib/public/AppFramework/Db/Entity.php:100]: But why is the variable called `name`? (also notice the dollar sign in the variable name lol)

**Comment by AndyScherzinger** [comment]
@miaulalala do you have a rough number how much faster the overall API call will then be with this improvement?

**Comment by miaulalala** [comment]
I don't know - i can do some comparisons against stable30 and master with blackfire?

**Comment by AndyScherzinger** [comment]
Only if it is not too much work. Was just thinking about it, in case it would be worth mentioning with v31, than marketing would need some numbers / comparison.

**Comment by provokateurin** [comment]
Just claim the `up to 6x` :joy: Technically not wrong is the best way to be right

File: lib/public/AppFramework/Db/Entity.php (hunk 2 above)
Comment (come-nc) [reply lib/public/AppFramework/Db/Entity.php:100]: The $ sign is not part of the name this is a dynamic property get.
It just checks if current value is already the one we’re setting and bails out. $name is the name of the property.

Comment (provokateurin) [reply lib/public/AppFramework/Db/Entity.php:100]: Ah I see, I didn't know this! Well I did know that this stuff is possible, but the syntax is horrific and I didn't recognize it

Comment (kesselb) [reply lib/public/AppFramework/Db/Entity.php:100]: Is `$this->{$name}` less horrific? ;)

Comment (kesselb) [reply lib/public/AppFramework/Db/Entity.php:100]: Or we move the call to a method with a self explaining name like isPropertyDifferent($name, $args[0])

Comment (provokateurin) [reply lib/public/AppFramework/Db/Entity.php:100]: > Is $this->{$name} less horrific? ;)

At least more recognizable to me.

**Comment by blizzz** [comment]
The added return type hints in lib/public/AppFramework/Db/Entity.php might be breaking, cf. https://github.com/nextcloud/officeonline/pull/611. They are also not documented in https://docs.nextcloud.com/server/latest/developer_manual/app_publishing_maintenance/app_upgrade_guide/upgrade_to_31.html.

**Pull request merged**